src/OptionContainer.cpp

```diff
@@ -75,140 +70,167 @@ void OptionContainer::deletePlugins(std::deque<Plugin *> &list)
+bool OptionContainer::readConfFile(const char *filename, String &list_pwd) {
+    std::string linebuffer;
+    String temp; // for tempory conversion and storage
+    String now_pwd(list_pwd);
+    std::ifstream conffiles(filename, std::ios::in); // e2guardianfN.conf
+    if (!conffiles.good()) {
+        if (!is_daemonised) {
+            std::cerr << thread_id << "Error reading: " << filename << std::endl;
+        }
+        syslog(LOG_ERR, "Error reading %s", filename);
+        return false;
+    }
+    String base_dir(filename);
+    base_dir.baseDir();
+
+    while (!conffiles.eof()) {
+        getline(conffiles, linebuffer);
+        if (!conffiles.fail() && linebuffer.length() != 0) {
+            if (linebuffer[0] != '#') { // i.e. not commented out
+                temp = (char *) linebuffer.c_str();
+                if (temp.contains("#")) {
+                    temp = temp.before("#");
+                }
+                temp.removeWhiteSpace(); // get rid of spaces at end of line
+                while (temp.contains("__LISTDIR__")) {
+                    String temp2 = temp.before("__LISTDIR__");
+                    temp2 += now_pwd;
+                    temp2 += temp.after("__LISTDIR__");
+                    temp = temp2;
+                }
+                // deal with included files
+                if (temp.startsWith(".")) {
+                    temp = temp.after(".Include<").before(">");
+                    if (temp.length() > 0) {
+                        temp.fullPath(base_dir);
+                        if (!readConfFile(temp.toCharArray(), now_pwd)) {
+                            conffiles.close();
+                            return false;
                         }
-                    } else 	if ((log_location = findoptionS("loglocation")) == "") {
-                        log_location = __LOGLOCATION;
-                        log_location += "/access.log";
-                        log_syslog = false;
                     }
+                    String temp2 = temp.after(".Define LISTDIR <").before(">");
+                    if (temp2.length() > 0) {
+                        now_pwd = temp2;
+                        //if(!now_pwd.endsWith("/"))
+                        //    now_pwd += "/";
+                    }
+                    continue;
+                }
+                // append ,listdir=now_pwd if line contains a file path - so that now_pwd can be passed
+                // to list file handler so that it can honour __LISTDIR__ in Included listfiles
+                if (temp.contains("path=") && !temp.contains("listdir=")) {
+                    temp += ",listdir=";
+                    temp += now_pwd;
+                }
+                linebuffer = temp.toCharArray();
+                conffile.push_back(linebuffer); // stick option in deque
+            }
+        }
+    }
+    conffiles.close();
+    return true;
+}
+
+bool OptionContainer::read(std::string &filename, int type) {
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/manual/numsys/e2guardian/pullRequest?prid=6238856)